### DISCUSSION TITLE
Welcome to python-facebook Discussions!

### DISCUSSION THREAD

**Discussion by MerleLiuKun** [sns-sdks/python-facebook#83] [Show and tell]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MerleLiuKun** [comment]
oh, Amazing area.

**Reply by MerleLiuKun** [reply]
tests for reply

**Comment by carlboudreau007** [comment]
I was able to get this to work.  But I struggled with the page_id, when I put in the Name, Nuber, and number as a string it failed.  But when I used page_id = 'me' it worked,

```
from flask import Flask, request, jsonify # , send_from_directory, abort

#pip install python-facebook-api # from 2023
# https://pypi.org/project/python-facebook-api/
from pyfacebook import GraphAPI

@app.route('/post_facebook_message', methods=['POST'])
def post_facebook_message():
    # https://pypi.org/project/python-facebook-api/
    # https://github.com/sns-sdks/python-facebook/tree/master

    # # - New test page
    my_access_token = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' # https://developers.facebook.com/tools/explorer/
    my_page_id = 'me'
    my_app_id = 'xxxxxxxxxxxxxxx' # https://developers.facebook.com/apps/xxxxxxxxxxxxxxx/settings/basic/
    my_app_secret = 'xxxxxxxxxxxxxxx' # https://developers.facebook.com/apps/xxxxxxxxxxxxxxx/settings/basic/
    
    # https://facebook-sdk.readthedocs.io/en/latest/api.html#put-object
    # if you already have an access token, you can initial with it
    api = GraphAPI(access_token=my_access_token) # Working, but expect it to fail in a few days due to it timing out.

    # if you want to use app credentials to generate app token
    # api = GraphAPI(app_id=my_app_id, app_secret=my_app_secret)  # Working for the get objects below but doesn't work for the post code

    # my_app_secret2 = api.get_app_token(app_id=my_app_id, app_secret=my_app_secret) seems to pull something back but it dosen't work for the post below
    # print(my_app_secret2["access_token"])

    # test1 = api.get_object(object_id=my_user_id) # WORKED - Returned Mwgg Passport
    # test2 = api.get_object(object_id=my_app_id) # WORKED - returned mwgg-2 and link to app profile
    # test3 = api.get_object(object_id=my_page_id) # Fails.  Use "me" to use the default page id.


    res = api.post_object(my_page_id,
        connection="feed",
        params={
            "fields": "id,message,created_time,from",
        },
        data={"message": "This is a test message by Flask API."},
    )

    # return (res['id'] + ' - ' + res['created_time']), 200
    return str(res), 200 # Returns the JSON string

```

**Reply by georgialive** [reply]
It looks like you're working on a Flask application to post messages to Facebook using the Python Facebook API (`pyfacebook`). Your code seems well-structured for this purpose. Here are a few points and suggestions based on what you've written:

1. **Access Token and Page ID**: You've used `'me'` as the `my_page_id`, which works when the access token is for a user. The `'me'` alias refers to the current user, so it defaults to their page. If you're using a page-specific access token, you should use the actual Page ID.

2. **App Credentials and Access Token**: It's good that you've commented out the alternative method of initializing the API with app credentials. Using an app token can be limiting for certain actions. For posting on a page, a user access token with the necessary permissions is generally required.

3. **Access Token Expiry**: You mentioned that the access token might time out. Typically, user access tokens have a short lifespan. You can either extend this token or implement a system to refresh it regularly.

4. **Testing and Debugging**: It's great that you've tested fetching objects using `api.get_object()`. For the post functionality, ensure that the token has the required permissions (`publish_pages` and `manage_pages`).

5. **Error Handling**: It might be beneficial to add error handling to manage and log API call failures. This can help in debugging and maintaining the application.

6. **Security Considerations**: Be cautious with your access tokens and app secrets. Ensure they're not exposed, especially in a public repository or in your codebase.

7. **Response and Status Code**: Your return statement sends back the API response as a string. This is fine for debugging but consider formatting this response to be more user-friendly or informative for the end-user or for logging purposes.

8. **Testing Posting Functionality**: You can test the posting functionality in a controlled environment, possibly using a test Facebook page, to ensure everything works as expected before deploying.

Remember, when dealing with APIs and external services, always handle sensitive data securely and comply with the terms of service of the API provider.

**Reply by MerleLiuKun** [reply]
hi. where your get this `page_id`

**Comment by carlboudreau007** [comment]
in the [README.rst](https://github.com/sns-sdks/python-facebook/tree/master#readme) 
The lines below do not work.  fb is undefined. If I tried to replace fb with api it failed because there is no method under api.user

```
Get user data:
>>> fb.user.get_info(user_id="413140042878187")
>>> User(id='413140042878187', name='Kun Liu')

Get page data:
>>> fb.page.get_info(page_id="20531316728")
>>> Page(id='20531316728', name='Facebook App')
```

**Reply by georgialive** [reply]
It seems like the issue you're encountering is due to incorrect usage of the `pyfacebook` library's API. The lines in the `README.rst` you're referring to probably assume a certain context or setup that isn't clearly mentioned. Let's address the problems:

1. **Undefined `fb` Variable**: The `fb` variable is likely intended to be an instance of the Facebook API, but it's not defined in the provided code snippet. This is probably a shorthand used in the documentation.

2. **No `user` Method in `api`**: The `api` instance you created with `GraphAPI` doesn't have a `user` method because `GraphAPI` is a more general class. The `pyfacebook` library might have different classes or methods for accessing user data, separate from the generic API class.

To address these issues:

- **Check the Library Documentation**: Make sure you're using the latest version of the `pyfacebook` library and refer to its official documentation. The methods and classes might differ between versions.

- **Correct Class or Method**: You might need to use a different class or method to access user data. The `GraphAPI` class might not be the correct one for this specific use case. Look for classes or methods specifically designed to interact with user or page data.

- **Example Adjustment**: If `fb` is intended to be an instance of a Facebook API class, you should create it accordingly. For example:
  
  ```python
  from pyfacebook import Api
  fb = Api(app_id="your_app_id", app_secret="your_app_secret", long_term_token="your_long_term_token")

  # Now use fb to get user or page information
  user_info = fb.user.get_info(user_id="413140042878187")
  page_info = fb.page.get_info(page_id="20531316728")
  ```

- **Use Correct Methods**: Depending on the version of `pyfacebook` you are using, the method to get user or page info might differ. Ensure you are using the correct method names as per the library's documentation.

- **Handling Library Updates**: If the library has been updated and the methods have changed, you might need to adjust your code accordingly. Sometimes, methods get deprecated or replaced in newer versions.

It's crucial to consult the official documentation of the library for the most accurate and up-to-date information on its usage.

**Reply by MerleLiuKun** [reply]
`fb` is not an instance for `GraphAPI` , It's `FacebookAPI` a subclass for `GraphAPI`
